File: code/game/WorldTile.cpp
#include "WorldTile.hpp"
#include "World.hpp"
#include "City.hpp"
#include "BaalExceptions.hpp"
#include "Geology.hpp"
#include "Weather.hpp"
#include "Engine.hpp"
#include "PlayerAI.hpp"

namespace baal {

///////////////////////////////////////////////////////////////////////////////
Yield::Yield(float food, float prod)
///////////////////////////////////////////////////////////////////////////////
  : m_food(food),
    m_prod(prod)
{
  // Tiles should either be food-yielding or production-yielding, not both
  Require(food == 0 || prod == 0, "Tile cannot yield food and production");
  Require(food >= 0, "Cannot have negative yields");
  Require(prod >= 0, "Cannot have negative yields");
}

///////////////////////////////////////////////////////////////////////////////
xmlNodePtr Yield::to_xml()
///////////////////////////////////////////////////////////////////////////////
{
  xmlNodePtr Yield_node = xmlNewNode(nullptr, BAD_CAST "Yield");
  std::ostringstream m_food_oss, m_prod_oss;
  m_food_oss << m_food;
  m_prod_oss << m_prod;
  xmlNewChild(Yield_node, nullptr, BAD_CAST "m_food", BAD_CAST m_food_oss.str().c_str());
  xmlNewChild(Yield_node, nullptr, BAD_CAST "m_prod", BAD_CAST m_prod_oss.str().c_str());

  return Yield_node;
}
/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
WorldTile::WorldTile(Location location, Yield yield, Climate& climate, Geology& geology)
///////////////////////////////////////////////////////////////////////////////
  : m_location(location),
    m_base_yield(yield),
    m_climate(climate),
    m_geology(geology),
    m_atmosphere(climate),
    m_worked(false),
    m_casted_spells()
{}

///////////////////////////////////////////////////////////////////////////////
WorldTile::~WorldTile()
///////////////////////////////////////////////////////////////////////////////
{
  delete &m_climate;
  delete &m_geology;
}

///////////////////////////////////////////////////////////////////////////////
void WorldTile::cycle_turn(const std::vector<std::shared_ptr<const Anomaly>>& anomalies,
                           const Location& location,
                           Season season)
///////////////////////////////////////////////////////////////////////////////
{
  m_geology.cycle_turn();
  m_atmosphere.cycle_turn(anomalies, location, season);
  m_worked = false;
  m_casted_spells.clear();
}

///////////////////////////////////////////////////////////////////////////////
void WorldTile::work()
///////////////////////////////////////////////////////////////////////////////
{
  Require(!m_worked, "Tile already being worked");
  m_worked = true;
}

///////////////////////////////////////////////////////////////////////////////
void WorldTile::cast(const std::string& spell)
///////////////////////////////////////////////////////////////////////////////
{
  Require(!contains(m_casted_spells, spell), "Duplicate: " << spell);
  m_casted_spells.push_back(spell);
}

///////////////////////////////////////////////////////////////////////////////
bool WorldTile::already_casted(const std::string& spell) const
///////////////////////////////////////////////////////////////////////////////
{
  return contains(m_casted_spells, spell);
}

///////////////////////////////////////////////////////////////////////////////
xmlNodePtr WorldTile::to_xml()
///////////////////////////////////////////////////////////////////////////////
{
  xmlNodePtr WorldTile_node = xmlNewNode(nullptr, BAD_CAST "Tile");

  xmlAddChild(WorldTile_node, m_base_yield.to_xml());
  xmlAddChild(WorldTile_node, m_climate.to_xml());
  xmlAddChild(WorldTile_node, m_geology.to_xml());
  xmlAddChild(WorldTile_node, m_atmosphere.to_xml());

  std::ostringstream worked_oss;
  worked_oss << m_worked;
  xmlNewChild(WorldTile_node, nullptr, BAD_CAST "m_worked", BAD_CAST worked_oss.str().c_str());

  return WorldTile_node;
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
OceanTile::OceanTile(Location location, unsigned depth, Climate& climate, Geology& geology) :
///////////////////////////////////////////////////////////////////////////////
  WorldTile(location, Yield(OCEAN_FOOD, OCEAN_PROD), climate, geology),
  m_depth(depth),
  m_surface_temp(0)
{
  // Current temp will be avg over all seasons
  for (Season s : iterate<Season>()) {
    m_surface_temp += climate.temperature(s);
  }
  m_surface_temp /= size<Season>();
}

///////////////////////////////////////////////////////////////////////////////
void OceanTile::cycle_turn(const std::vector<std::shared_ptr<const Anomaly>>& anomalies,
                           const Location& location,
                           Season season)
///////////////////////////////////////////////////////////////////////////////
{
  WorldTile::cycle_turn(anomalies, location, season);

  // Sea temperatures retain some heat, so new sea temps have to take old
  // sea temps into account. Here, we just average season temp and prior
  // sea temp together as a very simple model.
  m_surface_temp = new_surface_temp_func(m_surface_temp, m_atmosphere.temperature());
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
LandTile::LandTile(Location location, unsigned elevation, Yield yield, Climate& climate, Geology& geology)
///////////////////////////////////////////////////////////////////////////////
  : WorldTile(location, yield, climate, geology),
    m_hp(1.0),
    m_infra_level(0),
    m_elevation(elevation),
    m_snowpack(0), // TODO: Should have prexisting snowpack
    m_city(nullptr)
{}

///////////////////////////////////////////////////////////////////////////////
LandTile::~LandTile()
///////////////////////////////////////////////////////////////////////////////
{
  if (m_city) {
    delete m_city;
  }
}

///////////////////////////////////////////////////////////////////////////////
void LandTile::damage(float dmg)
///////////////////////////////////////////////////////////////////////////////
{
  Require(dmg >= 0.0 && dmg <= 1.0, "Invalid value for damage: " << dmg);

  m_hp *= (1.0 - dmg);

  Require(m_hp >= 0.0 && m_hp <= 1.0, "Invariant for hp failed: " << m_hp);
}

///////////////////////////////////////////////////////////////////////////////
void LandTile::cycle_turn(const std::vector<std::shared_ptr<const Anomaly>>& anomalies,
                          const Location& location,
                          Season season)
///////////////////////////////////////////////////////////////////////////////
{
  WorldTile::cycle_turn(anomalies, location, season);

  // Compute HP recovery
  m_hp = land_tile_recovery_func(m_hp);

  // Compute change in snowpack
  const float precip = atmosphere().precip();
  const int   temp   = atmosphere().temperature();

  // TODO - Take elevation into account

  const float snowfall_portion = portion_of_precip_that_falls_as_snow_func(temp);
  const float snowpack_melt_portion = portion_of_snowpack_that_melted(temp);
  const unsigned snowfall = (precip * 12) * snowfall_portion; // 12 inches of snow per liquid inch

  m_snowpack = (snowfall + m_snowpack) * (1 - snowpack_melt_portion);
}

///////////////////////////////////////////////////////////////////////////////
void LandTile::build_infra()
///////////////////////////////////////////////////////////////////////////////
{
  Require(m_infra_level < LAND_TILE_MAX_INFRA, "Infra is maxed");
  Require(city() == nullptr, "Cannot build infra if there is city here");

  m_infra_level++;
}

///////////////////////////////////////////////////////////////////////////////
void LandTile::destroy_infra(unsigned num_destroyed)
///////////////////////////////////////////////////////////////////////////////
{
  Require(m_infra_level >= num_destroyed, "num_destroyed too high");

  m_infra_level -= num_destroyed;
}

///////////////////////////////////////////////////////////////////////////////
Yield LandTile::yield() const
///////////////////////////////////////////////////////////////////////////////
{
  return compute_yield_func(m_base_yield, m_infra_level, m_hp);
}

///////////////////////////////////////////////////////////////////////////////
void LandTile::place_city(City& city)
///////////////////////////////////////////////////////////////////////////////
{
  Require(infra_level() == 0, "Cannot put city on tile with infra");
  Require(supports_city(), "Tile does not support cities");
  Require(m_city == nullptr, "Tile already had city: " << city.name());
  m_city = &city;
}

///////////////////////////////////////////////////////////////////////////////
void LandTile::remove_city()
///////////////////////////////////////////////////////////////////////////////
{
  Require(m_city != nullptr, "Erroneous call to remove_city");
  m_city = nullptr;
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
void MountainTile::place_city(City& city)
///////////////////////////////////////////////////////////////////////////////
{
  Require(false, "Mountain tiles cannot support cities");
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
void TileWithSoil::cycle_turn(const std::vector<std::shared_ptr<const Anomaly>>& anomalies,
                              const Location& location,
                              Season season)
///////////////////////////////////////////////////////////////////////////////
{
  LandTile::cycle_turn(anomalies, location, season);

  // MODEL: Model how precip and temp changes soil moisture

  // Get the parameters we need to make the calculation
  const float precip         = m_atmosphere.precip();
  const int temp             = m_atmosphere.temperature();
  const float avg_precip     = m_climate.precip(season);
  const int avg_temp         = m_climate.temperature(season);
  const float prior_moisture = m_soil_moisture;

  // Precip's effect on moisture
  const float precip_effect = precip_effect_on_moisture_func(avg_precip, precip);

  // Temperature's effect on moisture. This is trickier; for now only a very
  // basic model is in place.
  const float temp_effect = temp_effect_on_moisture_func(avg_temp, temp);

  // Compute overall seasonal forcing on soil moisture
  const float current_forcing = precip_effect * temp_effect;

  m_soil_moisture = compute_moisture_func(prior_moisture, current_forcing);

  Require(m_soil_moisture >= 0.0, "Moisture " << m_soil_moisture << " not valid");
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
Yield FoodTile::yield() const
///////////////////////////////////////////////////////////////////////////////
{
  Require (m_base_yield.m_food > 0, "Tiles with growth should yield food");

  // MODEL: Important equation that may need balancing

  // Food yielding tiles need to take soil moisture into account
  const float moisture_effect = moisture_yield_effect_func(m_soil_moisture);
  const float snowpack_effect = snowpack_yield_effect_func(m_snowpack);
  return LandTile::yield() * moisture_effect * snowpack_effect;
}

}

File: code/game/world_tile.py
#! /usr/bin/env python

"""
We put all WorldTile classes in this header to avoid
generating a ton of header files.
"""

import unittest

from baal_common import prequire, ProgramError, \
    check_access, check_callers, grant_access
from weather import Atmosphere, is_atmospheric, Climate
from geology import is_geological, Geology

###############################################################################
class Yield(object):
###############################################################################
    """
    A simple structure that specifies yields for tiles. You can think of this
    class as a named pair.
    """

    #
    # ==== Public API ====
    #

    def __init__(self, food, prod):
        prequire(food == 0 or prod == 0, "Tile cannot yield food and prod")
        prequire(food >= 0, "Cannot have negative yields")
        prequire(prod >= 0, "Cannot have negative yields")
        self.food = food
        self.prod = prod

    def __mul__(self, multiplier):
        return Yield(self.food * multiplier, self.prod * multiplier)

    def to_xml(self):
        # TODO - Aaron
        pass

# TODO - Support named properties instead of having individual getters/setters
# for each property?

###############################################################################
class WorldTile(object):
###############################################################################
    """
    WorldTile is the abstract base class of all tiles.

    Every tile has an atmosphere, climate, geology, and yield.

    This class also serves as an aggregate of all the *query*
    interfaces of the Tile subclasses. Anything you can possibly ask of
    any WorldTile must be associated with some method in the WorldTile
    class. None will be returned for non-sensical queries such as asking
    a mountain tile what its ocean-depth is.

    TODO: There is probably a better way to design this... consider not
    relying so heavily on inheritence.
    """

    #
    # ==== Public API ====
    #

    def __init__(self, yield_, climate, geology, location):
        self.__init_impl(yield_, climate, geology, location)

    #
    # Query/getter API
    #

    def location(self): return self.__location

    def yield_(self): return self.__yield_impl()

    def worked(self): return self.__worked

    def already_casted(self, spell): return spell.name() in self.__casted_spells

    def atmosphere(self): return self.__atmosphere

    def climate(self): return self.__climate

    def geology(self): return self.__geology

    def to_xml(self): return self.__to_xml_impl()

    def infra_level(self): return None

    def can_build_infra(self): return False

    def city(self): return None

    def supports_city(self): return False

    def soil_moisture(self): return None

    def depth(self): return None

    def sea_surface_temp(self): return None

    def elevation(self): return None

    def snowpack(self): return None

    #
    # Modification API
    #

    def work(self):
        """
        Mark this tile as being worked by a civilian
        """
        return self.__work_impl()

    def cycle_turn(self, anomalies, season):
        """
        Notify this tile that the turn is cycling, providing current
        anomalies and season so that we know how to adjust our atmosphere.
        """
        return self.__cycle_turn_impl(anomalies, season)

    def register_casted_spell(self, spell):
        """
        Notify this tile that a spell has been cast on it. We forbid casting
        the same spell multiple times on the same tile.
        """
        return self.__register_casted_spell_impl(spell)

    #
    # ==== Class constants ====
    #

    ALLOW_WORK           = "_world_tile_allow_work"
    ALLOW_CYCLE_TURN     = "_world_tile_allow_cycle_turn"
    ALLOW_REGISTER_SPELL = "_world_tile_allow_register_spell"

    #
    # ==== Implementation ====
    #

    ###########################################################################
    def __init_impl(self, yield_, climate, geology, location):
    ###########################################################################
        self.__base_yield    = yield_
        self.__climate       = climate
        self.__geology       = geology
        self.__atmosphere    = Atmosphere(climate)
        self.__location      = location
        self.__worked        = False
        self.__casted_spells = []

    ###########################################################################
    def __yield_impl(self):
    ###########################################################################
        from engine import engine

        # AI tech-level affects yield
        return engine().ai_player().get_adjusted_yield(self.__base_yield)

    ###########################################################################
    def __to_xml_impl(self):
    ###########################################################################
        # TODO - Aaron
        pass

    ###########################################################################
    def __work_impl(self):
    ###########################################################################
        check_access(self.ALLOW_WORK)

        prequire(not self.worked(), "Tile already being worked")
        self.__worked = True

    ###########################################################################
    def __cycle_turn_impl(self, anomalies, season):
    ###########################################################################
        check_access(self.ALLOW_CYCLE_TURN)

        self.__geology.cycle_turn()
        self.__atmosphere.cycle_turn(anomalies, self.location(), season)
        self.__worked = False

        self.__casted_spells = []

    ###########################################################################
    def __register_casted_spell_impl(self, spell):
    ###########################################################################
        check_access(self.ALLOW_REGISTER_SPELL)

        prequire(spell.name() not in self.__casted_spells)

        self.__casted_spells.append(spell.name())

grant_access(WorldTile, Atmosphere.ALLOW_CYCLE_TURN)
grant_access(WorldTile, Geology.ALLOW_CYCLE_TURN)

###############################################################################
class OceanTile(WorldTile):
###############################################################################
    """
    Represents an ocean tile. Ocean tiles have depth and surface temperature
    properties in addition to the normal set.
    """

    #
    # ==== Public API ====
    #

    def __init__(self, depth, climate, geology, location):
        self.__init_impl(depth, climate, geology, location)

    #
    # Queries/Getters
    #

    def depth(self): return self.__depth

    def sea_surface_temp(self): return self.__surface_temp

    #
    # Modification API
    #

    def cycle_turn(self, anomalies, season):
        return self.__cycle_turn_impl(anomalies, season)

    def set_surface_temp(self, new_temp):
        return self.__set_surface_temp_impl(new_temp)

    #
    # ==== Class Constants ====
    #

    ALLOW_SET_SURFACE_TEMP = "_ocean_tile_allow_set_surface_temp"
    _FOOD_YIELD = 3

    @classmethod
    def _NEW_SURFACE_TEMP_FUNC(cls, orig_surf_temp, new_air_temp):
        # Just average the two for now
        return (orig_surf_temp + new_air_temp) / 2

    #
    # ==== Implementation ====
    #

    ###########################################################################
    def __init_impl(self, depth, climate, geology, location):
    ###########################################################################
         super(OceanTile, self).__init__(Yield(self._FOOD_YIELD, 0),
                                         climate,
                                         geology,
                                         location)
         self.__depth = depth
         # TODO - Needs to be an average of all 4 seasons
         self.__surface_temp = self.atmosphere().temperature()

    ###########################################################################
    def __cycle_turn_impl(self, anomalies, season):
    ###########################################################################
        super(OceanTile, self).cycle_turn(anomalies, season)

        # Sea temperatures retain some heat, so new sea temps have to take old
        # sea temps into account. Here, we just average season temp and prior
        # sea temp together as a very simple model.
        self.__surface_temp = \
            self._NEW_SURFACE_TEMP_FUNC(self.__surface_temp,
                                        self.atmosphere().temperature())

    ###########################################################################
    def __set_surface_temp_impl(self, new_temp):
    ###########################################################################
        check_access(self.ALLOW_SET_SURFACE_TEMP)

        self.__surface_temp = new_temp

###############################################################################
class LandTile(WorldTile):
###############################################################################
    """
    Base class for all land tiles. Adds the concepts of tile damage, cities,
    infrastructure, elevation, and snowpack.
    """

    #
    # ==== Public API ====
    #

    def __init__(self, elevation, yield_, climate, geology, location):
        self.__init_impl(elevation, yield_, climate, geology, location)

    #
    # Queries/Getter API
    #

    def yield_(self): return self.__yield_impl()

    def infra_level(self): return self.__infra_level

    def city(self): return self.__city

    def supports_city(self): return True

    def can_build_infra(self):
        return self.infra_level() < self.LAND_TILE_MAX_INFRA

    def snowpack(self): return self.__snowpack

    def elevation(self): return self.__elevation

    #
    # Modification API
    #

    def damage(self, dmg): return self.__damage_impl(dmg)

    def cycle_turn(self, anomalies, season):
        return self.__cycle_turn_impl(anomalies, season)

    def build_infra(self): return self.__build_infra_impl()

    def destroy_infra(self, num_lvls_destroyed):
        return self.__destroy_infra_impl(num_lvls_destroyed)

    def place_city(self, city): return self.__place_city_impl(city)

    def remove_city(self): return self.__remove_city_impl()

    def set_snowpack(self, new_snowpack):
        return self.__set_snowpack_impl(new_snowpack)

    #
    # ==== Class Constants ====
    #

    LAND_TILE_MAX_INFRA = 5

    ALLOW_DAMAGE        = "_land_tile_damage"
    ALLOW_BUILD_INFRA   = "_land_tile_allow_build_infra"
    ALLOW_DESTROY_INFRA = "_land_tile_allow_destroy_infra"
    ALLOW_SET_SNOWPACK  = "_land_tile_allow_set_snowpack"

    #
    # Tweakable Constants
    #

    @classmethod
    def _LAND_TILE_RECOVERY_FUNC(cls, prior):
        # Recovers 10% of full hp per turn
        new_dmg = prior + .10
        return 1.0 if (new_dmg > 1.0) else new_dmg

    @classmethod
    def _COMPUTE_YIELD_FUNC(cls, normal_yield, infra_level, tile_hp):
        return normal_yield * (1 + infra_level) * tile_hp

    @classmethod
    def _PORTION_OF_PRECIP_THAT_FALLS_AS_SNOW_FUNC(cls, temp):
        if (temp < 30):
            return 1.0
        elif (temp < 60):
            return float(60 - temp) / 30
        else:
            return 0.0

    @classmethod
    def _PORTION_OF_SNOWPACK_THAT_MELTED(cls, temp):
        if (temp < 15):
            return 0.0
        elif (temp < 75):
            return float(temp - 15) / 60
        else:
            return 1.0

    #
    # ==== Implementation ====
    #

    ###########################################################################
    def __init_impl(self, elevation, yield_, climate, geology, location):
    ###########################################################################
        super(LandTile, self).__init__(yield_, climate, geology, location)

        self.__hp          = 1.0 # 0..1
        self.__infra_level = 0
        self.__city        = None
        self.__elevation   = elevation
        self.__snowpack    = 0 # TODO - Hack, should have preexiting snowpack

    ###########################################################################
    def __yield_impl(self):
    ###########################################################################
        normal_yield = super(LandTile, self).yield_()
        return self._COMPUTE_YIELD_FUNC(normal_yield,
                                        self.infra_level(),
                                        self.__hp)

    ###########################################################################
    def __damage_impl(self, dmg):
    ###########################################################################
        check_access(self.ALLOW_DAMAGE)

        prequire(dmg >= 0.0 and dmg <= 1.0, "Invalid ", dmg)

        self.__hp *= (1.0 - dmg)

        prequire(self.__hp >= 0.0 and self.__hp <= 1.0,
                 "Invariant failed ", self.__hp)

    ###########################################################################
    def __cycle_turn_impl(self, anomalies, season):
    ###########################################################################
        super(LandTile, self).cycle_turn(anomalies, season)

        # Compute HP recovery
        self.__hp = self._LAND_TILE_RECOVERY_FUNC(self.__hp)

        # Compute change in snowpack
        precip = self.atmosphere().precip()
        temp   = self.atmosphere().temperature()

        # TODO - Take elevation into account?

        snowfall_portion = \
            self._PORTION_OF_PRECIP_THAT_FALLS_AS_SNOW_FUNC(temp)
        snowpack_melt_portion = self._PORTION_OF_SNOWPACK_THAT_MELTED(temp)

        snowfall = (precip * 12) * snowfall_portion # 12 inches snow per inch

        self.__snowpack = (snowfall + self.snowpack()) * (1 - snowpack_melt_portion)

    ###########################################################################
    def __build_infra_impl(self):
    ###########################################################################
        check_access(self.ALLOW_BUILD_INFRA)

        prequire(self.infra_level() < self.LAND_TILE_MAX_INFRA, "max infra")
        prequire(self.city() is None, "Cannot build infra on city")

        self.__infra_level += 1

    ###########################################################################
    def __destroy_infra_impl(self, num_lvls_destroyed):
    ###########################################################################
        check_access(self.ALLOW_DESTROY_INFRA)

        prequire(self.infra_level() >= num_lvls_destroyed, "too high")

        self.__infra_level -= num_lvls_destroyed

    ###########################################################################
    def __place_city_impl(self, city):
    ###########################################################################
        check_callers(["place_city", "__place_city_impl_world"])

        prequire(self.infra_level() == 0, "Tile already has infra on it")
        prequire(self.supports_city(), "Tile type does not support cities")
        prequire(self.city() is None,  "Tile already has city")

        self.__city = city

    ###########################################################################
    def __remove_city_impl(self):
    ###########################################################################
        check_callers(["remove_city", "__remove_city_impl_world"])

        prequire(self.city() is not None, "No city to remove")

        self.__city = None

    ###########################################################################
    def __set_snowpack_impl(self, new_snowpack):
    ###########################################################################
        check_access(self.ALLOW_SET_SNOWPACK)

        self.__snowpack = new_snowpack

###############################################################################
class MountainTile(LandTile):
###############################################################################
    """
    Represents mountain tiles. Mountains don't add any new concepts. Cities
    can't be built on mountains.
    """

    def __init__(self, elevation, climate, geology, location):
        super(MountainTile, self).__init__(elevation,
                                           Yield(0, self._PROD_YIELD),
                                           climate,
                                           geology,
                                           location)

    def supports_city(self): return False

    _PROD_YIELD = 2

###############################################################################
class DesertTile(LandTile):
###############################################################################
    """
    Represents desert tiles. Deserts add no concepts, so this class is simple.
    """

    def __init__(self, elevation, climate, geology, location):
        super(DesertTile, self).__init__(elevation,
                                         Yield(0, self._PROD_YIELD),
                                         climate,
                                         geology,
                                         location)

    _PROD_YIELD = 0.5

###############################################################################
class TundraTile(LandTile):
###############################################################################
    """
    Represents tundra tiles. Tundra add no concepts, so this class is simple.
    """

    def __init__(self, elevation, climate, geology, location):
        super(TundraTile, self).__init__(elevation,
                                         Yield(0, self._PROD_YIELD),
                                         climate,
                                         geology,
                                         location)

    _PROD_YIELD = 0.5

###############################################################################
class HillsTile(LandTile):
###############################################################################
    """
    Represents hill tiles. Hills add no concepts, so this class is simple.

    TODO: Hill-tile lack of moisture make it awkward to handle
    for many disasters. Resolve this issue.
    """

    def __init__(self, elevation, climate, geology, location):
        super(HillsTile, self).__init__(elevation,
                                        Yield(0, self._PROD_YIELD),
                                        climate,
                                        geology,
                                        location)

    _PROD_YIELD = 1

###############################################################################
class FoodTile(LandTile):
###############################################################################
    """
    Represents land tiles that yield food. Introduces the concept of soil
    moisture.
    """

    def __init__(self, elevation, yield_, climate, geology, location):
        super(FoodTile, self).__init__(elevation, yield_, climate, geology, location)
        self.__soil_moisture = 1.0 # % of normal

    def soil_moisture(self): return self.__soil_moisture

    def yield_(self):
        normal_yield = super(FoodTile, self).yield_()
        return normal_yield * \
            self._MOISTURE_YIELD_EFFECT_FUNC(self.soil_moisture())

    def cycle_turn(self, anomalies, season):
        super(FoodTile, self).cycle_turn(anomalies, season)

        # Get the parameters we need to make the calculation
        precip         = self.atmosphere().precip()
        temp           = self.atmosphere().temperature()
        av_precip      = self.climate().precip(season)
        av_temp        = self.climate().temperature(season)
        prior_moisture = self.soil_moisture()

        # Precip's effect on moisture
        precip_effect = self._PRECIP_EFFECT_ON_MOISTURE_FUNC(av_precip, precip)

        # Temp's effect on moisture
        temp_effect = self._TEMP_EFFECT_ON_MOISTURE_FUNC(av_temp, temp)

        # Compute total effect
        total_recent_effect = precip_effect * temp_effect

        # Take past moisture into account but weight current moisture more heavily
        self.__soil_moisture = self._COMPUTE_MOISTURE(self.soil_moisture(),
                                                      total_recent_effect)

        prequire(self.__soil_moisture >= 0.0 and self.__soil_moisture < 100,
                "Moisture ", self.__soil_moisture, " not valid")

    FLOODING_THRESHOLD = 1.5
    TOTALLY_FLOODED    = 2.75

    @classmethod
    def _MOISTURE_YIELD_EFFECT_FUNC(cls, moisture):
        if (moisture < cls.FLOODING_THRESHOLD):
            # Up to the flooding threshold, yields improve as moisture
            # increases
            return moisture
        elif (moisture < cls.TOTALLY_FLOODED):
            # Yields drop quickly as soil becomes over-saturated
            return cls.FLOODING_THRESHOLD - (moisture - cls.FLOODING_THRESHOLD)
        else:
            # Things are flooded and can't get any worse. Farmers are able to
            # salvage some fixed portion of their crops.
            return 0.25

    @classmethod
    def _PRECIP_EFFECT_ON_MOISTURE_FUNC(cls, average_precip, precip):
        return float(precip) / average_precip

    @classmethod
    def _TEMP_EFFECT_ON_MOISTURE_FUNC(cls, average_temp, temp):
        return 1.0 + ( 0.01 * (average_temp - temp) )

    @classmethod
    def _COMPUTE_MOISTURE(cls, prior, current_forcing):
        return ((current_forcing * 2) + prior) / 3

###############################################################################
class PlainsTile(FoodTile):
###############################################################################
    """
    Represents plains tiles. Plains add no concepts, so this class is simple.
    """

    def __init__(self, elevation, climate, geology, location):
        super(PlainsTile, self).__init__(elevation,
                                         Yield(self._FOOD_YIELD, 0),
                                         climate,
                                         geology,
                                         location)

    _FOOD_YIELD = 1

###############################################################################
class LushTile(FoodTile):
###############################################################################
    """
    Represents lush tiles. Lush add no concepts, so this class is simple. Lush
    tiles are like plains tiles except they have higher food yields.
    """

    def __init__(self, elevation, climate, geology, location):
        super(LushTile, self).__init__(elevation,
                                       Yield(self._FOOD_YIELD, 0),
                                       climate,
                                       geology,
                                       location)

    _FOOD_YIELD = 2

#
# Tests
#

###############################################################################
class TestWorldTile(unittest.TestCase):
###############################################################################

    ###########################################################################
    def test_world_tile(self):
    ###########################################################################
        # TODO
        pass

if (__name__ == "__main__"):
    unittest.main()
